Stroop Updates - replace RandomizeColoredTextOrBox with Random Text Underlining

eschramm opened the pull request:
This PR looks to accomplish 2 goals:

1. After review with the client's psychologists, the intent here is to force "context switching". Switching between the box and the text does not do this. Instead, the recommendation is to randomly (50 %) underline the colored text and instruct the user to select the word displayed instead of the color of the word.

2. Allow for a specific test to be shown to the user for a practice test before the actual test - e.g., show one case of congruent color/text and then one that is incongruent, and finally one that is incongruent and underlined. This also will allow testing of functionality and results (see https://github.com/CareEvolution/CEVResearchKit/pull/204)

This should require no changes on the server to handle the updated StroopResult.

Tests added to ensure randomization works as expected.

eschramm committed:
add underline option for more context switching

the underline is intended to instruct the user to select the color word and not the font color

eschramm committed:
add breaks in ObjC switch statement

eschramm committed:
add underlined word example graphic

overcyn commented on ResearchKit/ActiveTasks/ORKStroopStepViewController.h:
```diff
@@ -39,6 +39,9 @@ NS_ASSUME_NONNULL_BEGIN
 @interface ORKStroopStepViewController : ORKActiveStepViewController
 
+// public for testing
+- (BOOL)randomBoolWithTrueProbability:(NSNumber * __nonnull)trueProbability;
```
Would this be a better fit in ORKHelpers_Internal.h?

eschramm replied:
Thanks for the suggestion. I really like the updated approach now which theoretically could be reused in other places in ResearchKit for randomization.

overcyn commented on ResearchKit/ActiveTasks/ORKStroopStep.m:
```diff
@@ -73,27 +117,43 @@ - (BOOL)startsFinished {
     self = [super initWithCoder:aDecoder];
     if (self ) {
         ORK_DECODE_INTEGER(aDecoder, numberOfAttempts);
+        ORK_DECODE_OBJ(aDecoder, probabilityOfVisualAndColorAlignment);
```
Does ORKStroopColor and ORKStroopTest need to implement NSCoder?

eschramm replied:
Hmm, probably. I see that we should also be adhering to `NSCopying` as well and have added that, too.

overcyn commented on ResearchKit/ActiveTasks/ORKStroopStep.m:
```diff
@@ -73,27 +117,43 @@ - (BOOL)startsFinished {
+            && [self.probabilityOfVisualAndColorAlignment isEqual:castObject.probabilityOfVisualAndColorAlignment]
+            && (self.stroopStyle == castObject.stroopStyle)
+            && (self.useGridLayoutForButtons == castObject.useGridLayoutForButtons)
+            && [self.nonRandomizedTests isEqual:castObject.nonRandomizedTests]);
```
Also does ORKStroopColor and ORKStroopTest need to implement `isEqual:`?

overcyn commented on ResearchKit/ActiveTasks/ORKStroopStep.h:
```diff
@@ -31,34 +31,82 @@
+ to provide a specific practice example set. If set, this overrides the randomizeVisualAndColorAlignment,
+ stroopStyle and numberOfAttempts properties.
+ */
+@property (nonatomic, strong, nullable) NSArray<ORKStroopTest *> *nonRandomizedTests;
```
NonRandomizedTests reads oddly to me (ie `stroopStep.nonRandomizedTests = [test1, test2, ...]`). What about just `stroopTests`?

eschramm replied:
Agreed, this sounds much cleaner as `stroopTests`.

overcyn reviewed (approved):
Looks really good. Minor comments inline.

eschramm committed:
PR improvements

- move randomBookWithTrueProbability: to inlined, global helper function
- rename nonRandomizedTests to stroopTests
- add NSSecureCoding, NSCopying adherance for new classes: ORKStroopColor, ORKStroopTest

overcyn commented:
LGTM
